File: src/src_core/fmg/core/mosaic/MosaicInitData.cs
using System;
using System.ComponentModel;
using fmg.common.geom;
using fmg.common.notifier;
using fmg.core.types;

namespace fmg.core.mosaic {

    /// <summary> Mosaic data </summary>
    public class MosaicInitData : INotifyPropertyChanged, IDisposable {

        public const double AREA_MINIMUM = 230;

        public const int MIN_SIZE_FIELD_M = 3;
        public const int MAX_SIZE_FIELD_M = 3000;
        public const int MIN_SIZE_FIELD_N = 3;
        public const int MAX_SIZE_FIELD_N = 3000;

        public const EMosaic       DEFAULT_MOSAIC_TYPE  = EMosaic.eMosaicSquare1;
        public const ESkillLevel   DEFAULT_SKILL_LEVEL  = ESkillLevel.eBeginner;
        public static readonly int DEFAULT_SIZE_FIELD_M = DEFAULT_SKILL_LEVEL.GetDefaultSize().m;
        public static readonly int DEFAULT_SIZE_FIELD_N = DEFAULT_SKILL_LEVEL.GetDefaultSize().n;
        public static readonly int DEFAULT_MINES_COUNT  = DEFAULT_SKILL_LEVEL.GetNumberMines(DEFAULT_MOSAIC_TYPE);

        private EMosaic   _mosaicType = DEFAULT_MOSAIC_TYPE;
        private Matrisize _sizeField  = new Matrisize(DEFAULT_SIZE_FIELD_M, DEFAULT_SIZE_FIELD_N);
        private int       _minesCount = DEFAULT_MINES_COUNT;

        private bool lockChanging = false;

        protected bool Disposed { get; private set; }
        private event PropertyChangedEventHandler PropertyChangedSync {
            add    { _notifier/*Sync*/.PropertyChanged += value;  }
            remove { _notifier/*Sync*/.PropertyChanged -= value;  }
        }
        public event PropertyChangedEventHandler PropertyChanged/*Async*/ {
            add    { _notifierAsync.PropertyChanged += value;  }
            remove { _notifierAsync.PropertyChanged -= value;  }
        }
        protected readonly NotifyPropertyChanged _notifier/*Sync*/;
        private   readonly NotifyPropertyChanged _notifierAsync;


        public MosaicInitData() {
            _notifier      = new NotifyPropertyChanged(this, false);
            _notifierAsync = new NotifyPropertyChanged(this, true);
            this.PropertyChangedSync += OnPropertyChanged;
        }


        public void CopyFrom(MosaicInitData from) {
            if (ReferenceEquals(from, this))
                return;
            this.MosaicType = from.MosaicType;
            this.SizeField  = from.SizeField;
            this.MinesCount = from.MinesCount;
        }

        public EMosaic MosaicType {
            get { return _mosaicType; }
            set { _notifier.SetProperty(ref _mosaicType, value); }
        }

        public EMosaicGroup MosaicGroup {
            get { return _mosaicType.GetGroup(); }
            set {
                if (_mosaicType.GetGroup() == value)
                    return;
                var ordinalInOldGroup = _mosaicType.GetOrdinalInGroup();
                var ordinalInNewGroup = Math.Min(ordinalInOldGroup, value.GetMosaics().Count - 1);
                MosaicType = value.GetMosaics()[ordinalInNewGroup];
            }
        }

        public Matrisize SizeField {
            get { return _sizeField; }
            set {
                if (value.m < MIN_SIZE_FIELD_M)
                    throw new ArgumentException("Size field M must be larger " + MIN_SIZE_FIELD_M);
                if (value.n < MIN_SIZE_FIELD_N)
                    throw new ArgumentException("Size field N must be larger " + MIN_SIZE_FIELD_N);
                if (value.m > MAX_SIZE_FIELD_M)
                    throw new ArgumentException("Size field M must be less " + (MAX_SIZE_FIELD_M + 1));
                if (value.n > MAX_SIZE_FIELD_N)
                    throw new ArgumentException("Size field N must be less " + (MAX_SIZE_FIELD_N + 1));
                _notifier.SetProperty(ref _sizeField, value);
            }
        }

        public int MinesCount {
            get { return _minesCount; }
            set { _notifier.SetProperty(ref _minesCount, value); }
        }

        public ESkillLevel SkillLevel {
            get {
                return ESkillLevelEx.CalcSkillLevel(MosaicType, SizeField, MinesCount);
            }
            set {
                if (value == ESkillLevel.eCustom)
                    throw new ArgumentException("Custom skill level not recognized");
                if (lockChanging)
                    throw new InvalidOperationException("Illegal usage");

                var skillOld = SkillLevel;
                if (skillOld == value)
                    return;

                lockChanging = true;
                try {
                    MinesCount = value.GetNumberMines(MosaicType);
                    SizeField = value.GetDefaultSize();
                } finally {
                    lockChanging = false;
                }

                var skillNew = SkillLevel;
                System.Diagnostics.Debug.Assert(value == skillNew);
                System.Diagnostics.Debug.Assert(value != skillOld);
                _notifier.FirePropertyChanged(skillOld, skillNew, nameof(SkillLevel));
            }
        }

        protected virtual void OnPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            // refire as async event
            _notifierAsync.FirePropertyChanged(ev);

            if (lockChanging)
                return;
            lockChanging = true;
            try {
                switch(ev.PropertyName) {
                case nameof(this.MosaicType):
                    {
                        var old = (ev as PropertyChangedExEventArgs<EMosaic>).OldValue;
                        if (old.GetGroup() != MosaicType.GetGroup())
                            _notifier.FirePropertyChanged(old.GetGroup(), MosaicType.GetGroup(), nameof(this.MosaicGroup));

                        var skillOld = ESkillLevelEx.CalcSkillLevel(old, SizeField, MinesCount);
                        if (skillOld == ESkillLevel.eCustom) {
                            var skillNew = SkillLevel;
                            if (skillNew != skillOld)
                                _notifier.FirePropertyChanged(skillOld, skillNew, nameof(this.SkillLevel));
                        } else {
                            // restore mines count for new mosaic type
                            MinesCount = skillOld.GetNumberMines(MosaicType);
                        }
                    }
                    break;
                case nameof(this.SizeField):
                    {
                        var skillOld = ESkillLevelEx.CalcSkillLevel(MosaicType, (ev as PropertyChangedExEventArgs<Matrisize>).OldValue, MinesCount);
                        var skillNew = SkillLevel;
                        if (skillNew != skillOld)
                            _notifier.FirePropertyChanged(skillOld, skillNew, nameof(this.SkillLevel));
                    }
                    break;
                case nameof(this.MinesCount):
                    {
                        var skillOld = ESkillLevelEx.CalcSkillLevel(MosaicType, SizeField, (ev as PropertyChangedExEventArgs<int>).OldValue);
                        var skillNew = SkillLevel;
                        if (skillNew != skillOld)
                            _notifier.FirePropertyChanged(skillOld, skillNew, nameof(this.SkillLevel));
                    }
                    break;
                }
            } finally {
                lockChanging = false;
            }
        }

        public override string ToString() {
            return nameof(MosaicInitData)
                + "{MosaicType:" + _mosaicType
                + ", SizeField:" + _sizeField
                + ", MinesCount:" + _minesCount
                + ", SkillLevel:" + SkillLevel
                + "}";
        }

        /// <summary>  Dispose managed resources </summary>/
        protected virtual void Disposing() {
            this.PropertyChangedSync -= OnPropertyChanged;
            _notifier.Dispose();
            _notifierAsync.Dispose();
        }

        public void Dispose() {
            if (Disposed)
                return;
            Disposed = true;
            Disposing();
            GC.SuppressFinalize(this);
        }

    }

}

File: src/src_core/fmg/core/mosaic/MosaicInitData.java
package fmg.core.mosaic;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import fmg.common.geom.Matrisize;
import fmg.common.notyfier.INotifyPropertyChanged;
import fmg.common.notyfier.NotifyPropertyChanged;
import fmg.core.types.EMosaic;
import fmg.core.types.ESkillLevel;

/** Mosaic data */
public class MosaicInitData implements INotifyPropertyChanged, AutoCloseable {

    public static final double AREA_MINIMUM = 230;

    public static final int MIN_SIZE_FIELD_M = 3;
    public static final int MAX_SIZE_FIELD_M = 3000;
    public static final int MIN_SIZE_FIELD_N = 3;
    public static final int MAX_SIZE_FIELD_N = 3000;

    public static final EMosaic     DEFAULT_MOSAIC_TYPE  = EMosaic.eMosaicSquare1;
    public static final ESkillLevel DEFAULT_SKILL_LEVEL  = ESkillLevel.eBeginner;
    public static final int         DEFAULT_SIZE_FIELD_M = DEFAULT_SKILL_LEVEL.getDefaultSize().m;
    public static final int         DEFAULT_SIZE_FIELD_N = DEFAULT_SKILL_LEVEL.getDefaultSize().n;
    public static final int         DEFAULT_MINES_COUNT  = DEFAULT_SKILL_LEVEL.getNumberMines(DEFAULT_MOSAIC_TYPE);

    private EMosaic mosaicType;
    private Matrisize sizeField;
    private int minesCount;

    private boolean lockChanging = false;

    protected final NotifyPropertyChanged _notifier/*Sync*/ = new NotifyPropertyChanged(this, false);
    private   final NotifyPropertyChanged _notifierAsync    = new NotifyPropertyChanged(this, true);

    public static final String PROPERTY_MOSAIC_TYPE = "MosaicType";
    public static final String PROPERTY_SIZE_FIELD  = "SizeField";
    public static final String PROPERTY_MINES_COUNT = "MinesCount";
    public static final String PROPERTY_SKILL_LEVEL = "SkillLevel";

    public MosaicInitData() {
        mosaicType = DEFAULT_MOSAIC_TYPE;
        sizeField  = new Matrisize(DEFAULT_SIZE_FIELD_M, DEFAULT_SIZE_FIELD_N);
        minesCount = DEFAULT_MINES_COUNT;
        _notifier.addListener(this::onPropertyChanged);
    }

    public void copyFrom(MosaicInitData from) {
        if (from == this)
            return;
        setMosaicType(from.getMosaicType());
        setSizeField( from.getSizeField());
        setMinesCount(from.getMinesCount());
    }

    public EMosaic getMosaicType() { return mosaicType; }
    public void setMosaicType(EMosaic mosaicType) {
        if (mosaicType == null)
            throw new IllegalArgumentException("Mosaic type can not be null");
        _notifier.setProperty(this.mosaicType, mosaicType, PROPERTY_MOSAIC_TYPE);
    }

    public Matrisize getSizeField() { return sizeField; }
    public void setSizeField(Matrisize sizeField) {
        if (sizeField == null)
            throw new IllegalArgumentException("Size field can not be null");
        if (sizeField.m < MIN_SIZE_FIELD_M)
            throw new IllegalArgumentException("Size field M must be larger " + MIN_SIZE_FIELD_M);
        if (sizeField.n < MIN_SIZE_FIELD_N)
            throw new IllegalArgumentException("Size field N must be larger " + MIN_SIZE_FIELD_N);
        if (sizeField.m > MAX_SIZE_FIELD_M)
            throw new IllegalArgumentException("Size field M must be less "  + (MAX_SIZE_FIELD_M + 1));
        if (sizeField.n > MAX_SIZE_FIELD_N)
            throw new IllegalArgumentException("Size field N must be less "  + (MAX_SIZE_FIELD_N + 1));

        _notifier.setProperty(this.sizeField, new Matrisize(sizeField.m, sizeField.n), PROPERTY_SIZE_FIELD);
    }

    public int getMinesCount() { return minesCount; }
    public void setMinesCount(int minesCount) {
        if (minesCount <= 0)
            throw new IllegalArgumentException("Mines count must be positive");

        _notifier.setProperty(this.minesCount, minesCount, PROPERTY_MINES_COUNT);
    }

    public ESkillLevel getSkillLevel() {
        return ESkillLevel.calcSkillLevel(mosaicType, sizeField, minesCount);
    }

    public void setSkillLevel(ESkillLevel skill) {
        if (skill == null)
            throw new IllegalArgumentException("Mosaic skill level can not be null");
        if (lockChanging)
            throw new UnsupportedOperationException("Illegal usage");

        ESkillLevel skillOld = getSkillLevel();
        if (skillOld == skill)
            return;

        lockChanging = true;
        try {
            setMinesCount(skill.getNumberMines(getMosaicType()));
            setSizeField(skill.getDefaultSize());
        } finally {
            lockChanging = false;
        }

        ESkillLevel skillNew = getSkillLevel();
        assert (skill == skillNew);
        assert (skill != skillOld);
        _notifier.firePropertyChanged(skillOld, skillNew, PROPERTY_SKILL_LEVEL);
    }

    protected void onPropertyChanged(PropertyChangeEvent ev) {
        // refire as async event
        _notifierAsync.firePropertyChanged(ev.getOldValue(), ev.getNewValue(), ev.getPropertyName());

        if (lockChanging)
            return;
        lockChanging = true;
        try {
            switch(ev.getPropertyName()) {
            case PROPERTY_MOSAIC_TYPE:
                {
                    ESkillLevel skillOld = ESkillLevel.calcSkillLevel((EMosaic)ev.getOldValue(), sizeField, minesCount);
                    if (skillOld == ESkillLevel.eCustom) {
                        ESkillLevel skillNew = getSkillLevel();
                        if (skillNew != skillOld)
                            _notifier.firePropertyChanged(skillOld, skillNew, PROPERTY_SKILL_LEVEL);
                    } else {
                        // restore mines count for new mosaic type
                        setMinesCount(skillOld.getNumberMines(getMosaicType()));
                    }
                }
                break;
            case PROPERTY_SIZE_FIELD:
                {
                    ESkillLevel skillOld = ESkillLevel.calcSkillLevel(mosaicType, (Matrisize)ev.getOldValue(), minesCount);
                    ESkillLevel skillNew = getSkillLevel();
                    if (!lockChanging && (skillNew != skillOld))
                        _notifier.firePropertyChanged(skillOld, skillNew, PROPERTY_SKILL_LEVEL);
                }
                break;
            case PROPERTY_MINES_COUNT:
                {
                    ESkillLevel skillOld = ESkillLevel.calcSkillLevel(mosaicType, sizeField, (int)ev.getOldValue());
                    ESkillLevel skillNew = getSkillLevel();
                    if (!lockChanging && (skillNew != skillOld))
                        _notifier.firePropertyChanged(skillOld, skillNew, PROPERTY_SKILL_LEVEL);
                }
                break;
            }
        } finally {
            lockChanging = false;
        }
    }

    @Override
    public void close() {
        _notifier.removeListener(this::onPropertyChanged);
        _notifier.close();
        _notifierAsync.close();
    }

    @Override
    public void addListener(PropertyChangeListener listener) {
        _notifierAsync.addListener(listener);
    }
    @Override
    public void removeListener(PropertyChangeListener listener) {
        _notifierAsync.removeListener(listener);
    }

}
